File: server/internal/clients/simplefin.go
package clients

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"
)

// SimpleFinConnection is the v2 connection object from the SimpleFIN bridge.
type SimpleFinConnection struct {
	ConnID  string `json:"conn_id"`
	Name    string `json:"name"`
	OrgID   string `json:"org_id"`
	OrgName string `json:"org_name"`
	OrgURL  string `json:"org_url"`
	SfinURL string `json:"sfin_url"`
}

// SimpleFinAccount is a single account from GET /accounts?version=2.
type SimpleFinAccount struct {
	ID               string                 `json:"id"`
	ConnID           string                 `json:"conn_id"`
	Name             string                 `json:"name"`
	Currency         string                 `json:"currency"`
	Balance          string                 `json:"balance"`
	AvailableBalance string                 `json:"available-balance"`
	BalanceDate      int64                  `json:"balance-date"`
	Transactions     []SimpleFinTransaction `json:"transactions"`
	Holdings         []SimpleFinHolding     `json:"holdings"`
	Extra            map[string]any         `json:"extra"`
}

// SimpleFinTransaction is one transaction from the SimpleFIN API.
type SimpleFinTransaction struct {
	ID           string         `json:"id"`
	Posted       int64          `json:"posted"`
	TransactedAt int64          `json:"transacted_at"`
	Amount       string         `json:"amount"`
	Description  string         `json:"description"`
	Payee        string         `json:"payee"`
	Memo         string         `json:"memo"`
	Pending      bool           `json:"pending"`
	Extra        map[string]any `json:"extra"`
}

// SimpleFinHolding is a portfolio holding (investment accounts only).
type SimpleFinHolding struct {
	ID            string `json:"id"`
	Symbol        string `json:"symbol"`
	Description   string `json:"description"`
	Shares        string `json:"shares"`
	MarketValue   string `json:"market_value"`
	CostBasis     string `json:"cost_basis"`
	PurchasePrice string `json:"purchase_price"`
	Currency      string `json:"currency"`
	Created       int64  `json:"created"`
}

// SimpleFinAccountSet is the v2 response from GET /accounts.
type SimpleFinAccountSet struct {
	Connections []SimpleFinConnection `json:"connections"`
	Accounts    []SimpleFinAccount    `json:"accounts"`
	Errors      []SimpleFinError      `json:"errlist"`
}

// SimpleFinError is a per-connection error from the SimpleFIN errlist.
type SimpleFinError struct {
	ConnID  string `json:"conn_id"`
	Message string `json:"message"`
}

// GetAccountsOpts carries optional parameters for GetAccounts.
type GetAccountsOpts struct {
	StartDate *time.Time
	Pending   bool
}

// SimpleFinClient is the interface for the SimpleFIN HTTP adapter.
type SimpleFinClient interface {
	// Claim decodes the Setup Token and POSTs to the claim URL to obtain the Access URL.
	Claim(ctx context.Context, setupToken string) (string, error)
	// GetAccounts calls GET /accounts?version=2 on the Access URL.
	GetAccounts(ctx context.Context, accessURL string, opts GetAccountsOpts) (SimpleFinAccountSet, error)
}

// SimpleFinHTTPClient is a stateless HTTP adapter for the SimpleFIN protocol.
type SimpleFinHTTPClient struct {
	HTTP *http.Client
}

var _ SimpleFinClient = (*SimpleFinHTTPClient)(nil)

func NewSimpleFinClient() *SimpleFinHTTPClient {
	return &SimpleFinHTTPClient{HTTP: &http.Client{Timeout: 30 * time.Second}}
}

// Claim decodes the base64 Setup Token to get the claim URL, then POSTs to it.
func (c *SimpleFinHTTPClient) Claim(ctx context.Context, setupToken string) (string, error) {
	claimURLBytes, err := base64.StdEncoding.DecodeString(strings.TrimSpace(setupToken))
	if err != nil {
		return "", fmt.Errorf("decode setup token: %w", err)
	}
	claimURL := strings.TrimSpace(string(claimURLBytes))
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, claimURL, http.NoBody)
	if err != nil {
		return "", fmt.Errorf("build claim request: %w", err)
	}
	resp, err := c.HTTP.Do(req)
	if err != nil {
		return "", fmt.Errorf("claim request: %w", err)
	}
	defer func() { _ = resp.Body.Close() }()
	if resp.StatusCode == http.StatusForbidden {
		return "", fmt.Errorf("setup token already claimed (HTTP 403)")
	}
	if resp.StatusCode != http.StatusOK {
		body, err := io.ReadAll(resp.Body)
		if err != nil {
			return "", fmt.Errorf("read claim error response: %w", err)
		}
		return "", fmt.Errorf("claim failed with HTTP %d: %s", resp.StatusCode, string(body))
	}
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("read claim response: %w", err)
	}
	accessURL := strings.TrimSpace(string(body))
	if accessURL == "" {
		return "", fmt.Errorf("empty access URL returned by claim endpoint")
	}
	if _, err := url.Parse(accessURL); err != nil {
		return "", errors.New("invalid simplefin access URL")
	}
	return accessURL, nil
}

// GetAccounts calls GET /accounts?version=2 on the Access URL.
func (c *SimpleFinHTTPClient) GetAccounts(ctx context.Context, accessURL string, opts GetAccountsOpts) (SimpleFinAccountSet, error) {
	endpoint, err := buildAccountsURL(accessURL, opts)
	if err != nil {
		return SimpleFinAccountSet{}, err
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, endpoint, http.NoBody)
	if err != nil {
		return SimpleFinAccountSet{}, fmt.Errorf("build accounts request: %w", err)
	}
	injectBasicAuth(req, accessURL)
	resp, err := c.HTTP.Do(req)
	if err != nil {
		return SimpleFinAccountSet{}, fmt.Errorf("accounts request: %w", err)
	}
	defer func() { _ = resp.Body.Close() }()
	if resp.StatusCode != http.StatusOK {
		body, err := io.ReadAll(resp.Body)
		if err != nil {
			return SimpleFinAccountSet{}, fmt.Errorf("read accounts error response: %w", err)
		}
		return SimpleFinAccountSet{}, fmt.Errorf("accounts request failed with HTTP %d: %s", resp.StatusCode, string(body))
	}
	var result SimpleFinAccountSet
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return SimpleFinAccountSet{}, fmt.Errorf("decode accounts response: %w", err)
	}
	if err := validateSimpleFinCurrencies(result); err != nil {
		return SimpleFinAccountSet{}, err
	}
	return result, nil
}

// buildAccountsURL appends the /accounts path and query params to the Access URL.
func buildAccountsURL(accessURL string, opts GetAccountsOpts) (string, error) {
	// Sanitized error: the raw URL can embed Basic Auth credentials and
	// url.Parse errors echo their input.
	u, err := url.Parse(accessURL)
	if err != nil {
		return "", errors.New("invalid simplefin access URL")
	}
	// Clear credentials from URL; they go in the Authorization header.
	u.User = nil
	// Ensure path ends with /accounts.
	if !strings.HasSuffix(u.Path, "/accounts") {
		u.Path = strings.TrimRight(u.Path, "/") + "/accounts"
	}
	q := u.Query()
	q.Set("version", "2")
	if opts.Pending {
		q.Set("pending", "1")
	}
	if opts.StartDate != nil {
		q.Set("start-date", fmt.Sprintf("%d", opts.StartDate.Unix()))
	}
	u.RawQuery = q.Encode()
	return u.String(), nil
}

func validateSimpleFinCurrencies(accountSet SimpleFinAccountSet) error {
	for _, account := range accountSet.Accounts {
		if account.Currency != "" && !strings.EqualFold(account.Currency, "USD") {
			return fmt.Errorf("unsupported simplefin account currency %q", account.Currency)
		}
		for _, holding := range account.Holdings {
			if holding.Currency != "" && !strings.EqualFold(holding.Currency, "USD") {
				return fmt.Errorf("unsupported simplefin holding currency %q", holding.Currency)
			}
		}
	}
	return nil
}

// injectBasicAuth extracts userinfo from the Access URL and sets the Authorization header.
func injectBasicAuth(req *http.Request, accessURL string) {
	u, err := url.Parse(accessURL)
	if err != nil || u.User == nil {
		return
	}
	user := u.User.Username()
	pass, _ := u.User.Password()
	req.SetBasicAuth(user, pass)
}

File: server-rs/src/clients/simplefin.rs
use anyhow::{Context, Result};
use base64::{Engine, engine::general_purpose::STANDARD};
use chrono::{DateTime, Utc};
use reqwest::Client;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use url::Url;

const TIMEOUT_SECONDS: u64 = 30;

#[derive(Clone, Debug, Default, Deserialize, PartialEq, Serialize)]
#[serde(default)]
pub struct SimpleFinConnection {
    pub conn_id: String,
    pub name: String,
    pub org_id: String,
    pub org_name: String,
    pub org_url: String,
    pub sfin_url: String,
}

#[derive(Clone, Debug, Default, Deserialize, PartialEq, Serialize)]
#[serde(default)]
pub struct SimpleFinAccount {
    pub id: String,
    pub conn_id: String,
    pub name: String,
    pub currency: String,
    pub balance: String,
    #[serde(rename = "available-balance")]
    pub available_balance: String,
    #[serde(rename = "balance-date")]
    pub balance_date: i64,
    pub transactions: Vec<SimpleFinTransaction>,
    pub holdings: Vec<SimpleFinHolding>,
    pub extra: serde_json::Map<String, Value>,
}

#[derive(Clone, Debug, Default, Deserialize, PartialEq, Serialize)]
#[serde(default)]
pub struct SimpleFinTransaction {
    pub id: String,
    pub posted: i64,
    pub transacted_at: i64,
    pub amount: String,
    pub description: String,
    pub payee: String,
    pub memo: String,
    pub pending: bool,
    pub extra: serde_json::Map<String, Value>,
}

#[derive(Clone, Debug, Default, Deserialize, PartialEq, Serialize)]
#[serde(default)]
pub struct SimpleFinHolding {
    pub id: String,
    pub symbol: String,
    pub description: String,
    pub shares: String,
    pub market_value: String,
    pub cost_basis: String,
    pub purchase_price: String,
    pub currency: String,
    pub created: i64,
}

#[derive(Clone, Debug, Default, Deserialize, PartialEq, Serialize)]
#[serde(default)]
pub struct SimpleFinAccountSet {
    pub connections: Vec<SimpleFinConnection>,
    pub accounts: Vec<SimpleFinAccount>,
    #[serde(rename = "errlist")]
    pub errors: Vec<SimpleFinError>,
}

#[derive(Clone, Debug, Default, Deserialize, PartialEq, Serialize)]
#[serde(default)]
pub struct SimpleFinError {
    pub conn_id: String,
    pub message: String,
}

#[derive(Clone, Debug, Default)]
pub struct GetAccountsOpts {
    pub start_date: Option<DateTime<Utc>>,
    pub pending: bool,
}

#[derive(Clone)]
pub struct SimpleFinClient {
    http: Client,
}

impl SimpleFinClient {
    pub fn new() -> Result<Self> {
        Ok(Self {
            http: Client::builder()
                .timeout(std::time::Duration::from_secs(TIMEOUT_SECONDS))
                .build()
                .context("build SimpleFIN HTTP client")?,
        })
    }

    pub async fn claim(&self, setup_token: &str) -> Result<String> {
        let claim_url = STANDARD.decode(setup_token.trim()).context("decode setup token")?;
        let claim_url = String::from_utf8(claim_url).context("decode setup token")?;
        let response = self.http.post(claim_url.trim()).send().await.context("claim request")?;
        let status = response.status();
        if status == reqwest::StatusCode::FORBIDDEN {
            anyhow::bail!("setup token already claimed (HTTP 403)");
        }
        let body = response.text().await.with_context(|| {
            if status == reqwest::StatusCode::OK { "read claim response" } else { "read claim error response" }
        })?;
        if status != reqwest::StatusCode::OK {
            anyhow::bail!("claim failed with HTTP {}: {body}", status.as_u16());
        }

        let access_url = body.trim().to_owned();
        anyhow::ensure!(!access_url.is_empty(), "empty access URL returned by claim endpoint");
        parse_access_url(&access_url)?;
        Ok(access_url)
    }

    pub async fn get_accounts(&self, access_url: &str, opts: GetAccountsOpts) -> Result<SimpleFinAccountSet> {
        let access_url = parse_access_url(access_url)?;
        let username = access_url.username().to_owned();
        let password = access_url.password().map(str::to_owned);
        let endpoint = accounts_url(access_url, opts)?;
        let request = if username.is_empty() {
            self.http.get(endpoint)
        } else {
            self.http.get(endpoint).basic_auth(username, password)
        };
        let response = request.send().await.context("accounts request")?;
        let status = response.status();
        let body = response.text().await.with_context(|| {
            if status == reqwest::StatusCode::OK { "read accounts response" } else { "read accounts error response" }
        })?;
        if status != reqwest::StatusCode::OK {
            anyhow::bail!("accounts request failed with HTTP {}: {body}", status.as_u16());
        }

        let accounts = serde_json::from_str::<SimpleFinAccountSet>(&body).context("decode accounts response")?;
        validate_currencies(&accounts)?;
        Ok(accounts)
    }
}

fn parse_access_url(access_url: &str) -> Result<Url> {
    Url::parse(access_url).map_err(|_| anyhow::anyhow!("invalid simplefin access URL"))
}

fn accounts_url(mut access_url: Url, opts: GetAccountsOpts) -> Result<Url> {
    access_url
        .set_username("")
        .map_err(|_| anyhow::anyhow!("invalid simplefin access URL"))?;
    access_url
        .set_password(None)
        .map_err(|_| anyhow::anyhow!("invalid simplefin access URL"))?;
    if !access_url.path().ends_with("/accounts") {
        let path = format!("{}/accounts", access_url.path().trim_end_matches('/'));
        access_url.set_path(&path);
    }
    let query_pairs = access_url
        .query_pairs()
        .filter(|(name, _)| name != "version" && name != "pending" && name != "start-date")
        .map(|(name, value)| (name.into_owned(), value.into_owned()))
        .collect::<Vec<_>>();
    {
        let mut query = access_url.query_pairs_mut();
        query.clear().extend_pairs(query_pairs);
        query.append_pair("version", "2");
        if opts.pending {
            query.append_pair("pending", "1");
        }
        if let Some(start_date) = opts.start_date {
            query.append_pair("start-date", &start_date.timestamp().to_string());
        }
    }
    Ok(access_url)
}

fn validate_currencies(accounts: &SimpleFinAccountSet) -> Result<()> {
    for account in &accounts.accounts {
        if !account.currency.is_empty() && !account.currency.eq_ignore_ascii_case("USD") {
            anyhow::bail!("unsupported simplefin account currency {:?}", account.currency);
        }
        for holding in &account.holdings {
            if !holding.currency.is_empty() && !holding.currency.eq_ignore_ascii_case("USD") {
                anyhow::bail!("unsupported simplefin holding currency {:?}", holding.currency);
            }
        }
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use base64::{Engine, engine::general_purpose::STANDARD};
    use chrono::{TimeZone, Utc};
    use wiremock::{
        Mock, MockServer, ResponseTemplate,
        matchers::{basic_auth, method, path, query_param},
    };

    use super::{GetAccountsOpts, SimpleFinClient};

    #[tokio::test]
    async fn claims_a_setup_token_and_fetches_accounts() {
        let server = MockServer::start().await;
        let access_url = format!(
            "{}://user:pass@{}/simplefin",
            server.uri().split_once("://").unwrap().0,
            server.address()
        );
        Mock::given(method("POST"))
            .and(path("/claim"))
            .respond_with(ResponseTemplate::new(200).set_body_string(access_url))
            .mount(&server)
            .await;
        Mock::given(method("GET"))
            .and(path("/simplefin/accounts"))
            .and(basic_auth("user", "pass"))
            .and(query_param("version", "2"))
            .and(query_param("pending", "1"))
            .and(query_param("start-date", "1780272000"))
            .respond_with(ResponseTemplate::new(200).set_body_string(ACCOUNTS))
            .mount(&server)
            .await;

        let client = SimpleFinClient::new().unwrap();
        let token = STANDARD.encode(format!("{}/claim", server.uri()));
        let access_url = client.claim(&token).await.unwrap();
        let accounts = client
            .get_accounts(
                &access_url,
                GetAccountsOpts {
                    start_date: Some(Utc.with_ymd_and_hms(2026, 6, 1, 0, 0, 0).unwrap()),
                    pending: true,
                },
            )
            .await
            .unwrap();

        assert_eq!(accounts.connections.len(), 1);
        assert_eq!(accounts.accounts[0].holdings[0].shares, "3539.578");
        assert_eq!(accounts.accounts[0].holdings[0].market_value, "100736.38");
        assert_eq!(accounts.accounts[0].holdings[0].cost_basis, "155506.10");
    }

    #[tokio::test]
    async fn rejects_invalid_tokens_and_non_usd_accounts() {
        let client = SimpleFinClient::new().unwrap();
        assert!(client.claim("not-base64").await.is_err());
        assert_eq!(
            client
                .get_accounts("http://user:secret@bad host/simplefin", GetAccountsOpts::default())
                .await
                .unwrap_err()
                .to_string(),
            "invalid simplefin access URL"
        );

        let server = MockServer::start().await;
        Mock::given(method("GET"))
            .respond_with(ResponseTemplate::new(200).set_body_string(r#"{"accounts":[{"currency":"EUR"}]}"#))
            .mount(&server)
            .await;
        assert!(
            client
                .get_accounts(&server.uri(), GetAccountsOpts::default())
                .await
                .is_err()
        );
    }

    const ACCOUNTS: &str = r#"{
      "connections": [{"conn_id": "conn-1", "name": "Bank"}],
      "accounts": [{
        "id": "acc",
        "conn_id": "conn-1",
        "name": "Checking",
        "holdings": [{
          "id": "holding-1",
          "symbol": "VGIT",
          "description": "Vanguard Intermediate-Term Treasury Index Fund",
          "shares": "3539.578",
          "market_value": "100736.38",
          "cost_basis": "155506.10",
          "purchase_price": "21.9506",
          "currency": "USD",
          "created": 1780000000
        }]
      }]
    }"#;
}
